File: TypeDefinitions/WorldObjectProxy.cs
namespace WorldServer.Scripting.MoonSharp.Proxies;

using Common.Positions;
using global::MoonSharp.Interpreter;
using JetBrains.Annotations;
using WorldServer.World.Objects;

[PublicAPI]
public class WorldObjectProxy
{
    private readonly WorldObject _worldObject;

    [MoonSharpHidden]
    public WorldObjectProxy(WorldObject worldObject)
    {
        _worldObject = worldObject;
    }

    [PublicAPI]
    public string Name => _worldObject.Name;

    [PublicAPI]
    public void PlaySound(ushort soundId) => _worldObject.PlaySound(soundId);

    [PublicAPI]
    public void PlayEffect(ushort effectId) => _worldObject.PlayEffect(effectId);

    [PublicAPI]
    public HashSet<Player> PlayersInRange => _worldObject.PlayersInRange;

    [PublicAPI]
    public HashSet<WorldObject> ObjectsInRange => _worldObject.ObjectsInRange;

    [PublicAPI]
    public List<Player> GetPlayersInRange(uint range = 2400) => _worldObject.PlayersInRange.Where(p => p.WorldPosition.IsWithin3DRadiusUnits(_worldObject.WorldPosition, range)).ToList();

    [PublicAPI]
    public List<GameObject> GetGameObjectsInRange(uint range = 2400, uint? gameObjectEntryId = null) => _worldObject.ObjectsInRange.Where(o => o is GameObject gameObject && (gameObjectEntryId is null || gameObject.Entry == gameObjectEntryId) && gameObject.WorldPosition.IsWithin3DRadiusUnits(_worldObject.WorldPosition, range)).OfType<GameObject>().ToList();

    [PublicAPI]
    public List<Creature> GetCreaturesInRange(uint range = 2400, uint? creatureEntryId = null) => _worldObject.ObjectsInRange.Where(o => o is Creature creature && (creatureEntryId is null || creature.Entry == creatureEntryId) && creature.WorldPosition.IsWithin3DRadiusUnits(_worldObject.WorldPosition, range)).OfType<Creature>().ToList();

    [PublicAPI]
    public GameObject? SpawnGameObject(uint gameObjectProtoId, uint x, uint y, ushort z, ushort o) => _worldObject.Region?.CreateGameObject(gameObjectProtoId, new Point3D(x, y, z), o) ?? null;

    [PublicAPI]
    public GameObject? GetNearestGameObject(uint range = 2400, uint? gameObjectEntryId = null) => _worldObject.ObjectsInRange.Where(o => o is GameObject gameObject && (gameObjectEntryId is null || gameObject.Entry == gameObjectEntryId) && gameObject.WorldPosition.IsWithin3DRadiusUnits(_worldObject.WorldPosition, range)).MinBy(o => o.Get3DDistanceUnits(_worldObject)) as GameObject;

    [PublicAPI]
    public Creature? GetNearestCreature(uint range = 2400, uint? creatureEntryId = null) => _worldObject.ObjectsInRange.Where(o => o is Creature creature && (creatureEntryId is null || creature.Entry == creatureEntryId) && creature.WorldPosition.IsWithin3DRadiusUnits(_worldObject.WorldPosition, range)).MinBy(o => o.Get3DDistanceUnits(_worldObject)) as Creature;

    [PublicAPI]
    public Player? GetNearestPlayer(uint range = 2400, uint? gameObjectEntryId = null) => _worldObject.PlayersInRange.Where(p => p.WorldPosition.IsWithin3DRadiusUnits(_worldObject.WorldPosition, range)).MinBy(p => p.Get3DDistanceUnits(_worldObject));

    [PublicAPI]
    public double GetDistanceTo(Unit other) => _worldObject.WorldPosition.Get3DDistanceUnits(other.WorldPosition);

    #region World

    [PublicAPI]
    public ushort RegionId => _worldObject.Region?.RegionId ?? 0;

    [PublicAPI]
    public ushort ZoneId => _worldObject.ZoneId;

    [PublicAPI]
    public uint X => _worldObject.WorldPosition.X;

    [PublicAPI]
    public uint Y => _worldObject.WorldPosition.Y;

    [PublicAPI]
    public uint Z => _worldObject.WorldPosition.Z;

    [PublicAPI]
    public uint O => _worldObject.Heading;

    [PublicAPI]
    public bool IsWithin3DRadiusFeet(WorldObject worldObject, uint range) => _worldObject.IsWithin3DRadiusFeet(worldObject, range);

    #endregion

    #region Is

    [PublicAPI]
    public bool IsPlayer => _worldObject is Player;

    [PublicAPI]
    public bool IsCreature => _worldObject is Creature;

    [PublicAPI]
    public bool IsGameObject => _worldObject is GameObject;

    [PublicAPI]
    public bool IsUnit => _worldObject is Unit;

    [PublicAPI]
    public Player? AsPlayer => _worldObject as Player;

    [PublicAPI]
    public Creature? AsCreature => _worldObject as Creature;

    [PublicAPI]
    public GameObject? AsGameObject => _worldObject as GameObject;

    [PublicAPI]
    public Unit? AsUnit => _worldObject as Unit;

    #endregion
}

File: TypeDefinitions/WorldObjectProxy.lua
-- world_object_proxy.lua

---@class WorldObjectProxy
---@field Name string
---@field PlayersInRange table<number, PlayerProxy>
---@field ObjectsInRange table<number, WorldObjectProxy>
---@field RegionId number
---@field ZoneId number
---@field X number
---@field Y number
---@field Z number
---@field O number
---@field IsPlayer boolean
---@field IsCreature boolean
---@field IsGameObject boolean
---@field IsUnit boolean
---@field AsPlayer PlayerProxy|nil
---@field AsCreature CreatureProxy|nil
---@field AsGameObject GameObjectProxy|nil
---@field AsUnit UnitProxy|nil
local WorldObjectProxy = {}

--- Plays a sound.
---@param soundId number
function WorldObjectProxy:PlaySound(soundId) end

--- Plays an effect.
---@param effectId number
function WorldObjectProxy:PlayEffect(effectId) end

--- Gets players in range within the specified distance.
---@param range number
---@return PlayerProxy[]
function WorldObjectProxy:GetPlayersInRange(range) end

--- Gets game objects in range within the specified distance.
---@param range number
---@param gameObjectEntryId number|nil
---@return GameObjectProxy[]
function WorldObjectProxy:GetGameObjectsInRange(range, gameObjectEntryId) end

--- Gets creatures in range within the specified distance.
---@param range number
---@param creatureEntryId number|nil
---@return CreatureProxy[]
function WorldObjectProxy:GetCreaturesInRange(range, creatureEntryId) end

--- Spawns a game object.
---@param gameObjectProtoId number
---@param x number
---@param y number
---@param z number
---@param o number
---@return GameObjectProxy|nil
function WorldObjectProxy:SpawnGameObject(gameObjectProtoId, x, y, z, o) end

--- Gets the nearest game object within the specified distance.
---@param range number
---@param gameObjectEntryId number|nil
---@return GameObjectProxy|nil
function WorldObjectProxy:GetNearestGameObject(range, gameObjectEntryId) end

--- Gets the nearest creature within the specified distance.
---@param range number
---@param creatureEntryId number|nil
---@return CreatureProxy|nil
function WorldObjectProxy:GetNearestCreature(range, creatureEntryId) end

--- Gets the nearest player within the specified distance.
---@param range number
---@param gameObjectEntryId number|nil
---@return PlayerProxy|nil
function WorldObjectProxy:GetNearestPlayer(range, gameObjectEntryId) end

--- Checks if the world object is within a 3D radius.
---@param worldObject WorldObjectProxy
---@param range number
---@return boolean
function WorldObjectProxy:IsWithin3DRadiusFeet(worldObject, range) end

return WorldObjectProxy
